llumnix/backends/vllm/migration_backend.py

```diff
@@ -0,0 +1,277 @@
+
+        self.migration_cache_size = self.cache_engine.block_size * self.cache_engine.num_heads * self.cache_engine.head_size
+
+        if self.backend == 'gloo':
+            try_import_gloo()
+            self.cache_device = "cpu"
+        elif self.backend == 'nccl':
+            self.cache_device = torch.device(f"cuda:{self.local_rank}")
+        else:
+            raise ValueError("backend must be 'gloo' or 'nccl'")
+
+        pin_memory = (self.backend == 'gloo')
+        self.dummy_cache = torch.empty(
+            size=(self.num_migration_cache_blocks, self.migration_num_layers, 2, self.migration_cache_size),
+            dtype=self.cache_engine.dtype,
+            device=self.cache_device,
+            pin_memory=pin_memory
+        )
+
+        self.migration_stream = torch.cuda.Stream()
+
+    def init_backend(self, group_name, world_size, rank) -> bool:
+        @func_set_timeout(self.migration_config.migration_backend_init_timeout)
+        def init_group(world_size, rank, backend, group_name):
+            col.init_collective_group(world_size, rank, backend, group_name)
+
+        try:
+            init_group(world_size, rank, self.backend, group_name)
+        except FunctionTimedOut:
+            logger.info("create ray collective group fail (group_name:{}, world_size: {}, rank: {}, backbend: {})."
+                .format(group_name, world_size, rank, self.backend))
+            return False
+
+        self.group_name = group_name
+        self.global_world_size = world_size
+        self.global_rank = rank
+
+        logger.info("create ray collective group successfully (group_name:{}, world_size: {}, rank: {}, backbend: {})."
+                    .format(self.group_name, self.global_world_size, self.global_rank, self.backend))
+        return True
+
+    def warmup(self):
+        if self.global_world_size > 1 and self.group_name is not None:
+            try:
+                col.allreduce(self.dummy_cache[0], self.group_name)
+            # pylint: disable=W0703
+            except Exception as e:
+                logger.info("warmup collective group failed (group_name:{}, world_size: {}, rank: {}, backbend: {}), err: {}."
+                    .format(self.group_name, self.global_world_size, self.global_rank, self.backend, e))
+                return False
+
+        logger.info("ray collective group warmup successfully (group_name:{}, world_size: {}, rank: {}, backbend: {})."
+                    .format(self.group_name, self.global_world_size, self.global_rank, self.backend))
+        return True
+
+    def destory_backend(self) -> None:
+        if self.group_name is None:
+            return
+
+        err_info = None
+        try:
+            col.destroy_collective_group(self.group_name)
+        # pylint: disable=W0703
+        except Exception as e:
+            err_info = e
+        logger.info("destory ray collective group successfully (group_name:{}, backbend: {}), meet_err: {}."
```

Comment:
sucessfully, but meet exception?  Also, do not output meet_err.

Reply:
get